// @keystone-next/admin-ui/dist/Errors-6f077908.cjs.dev.js
'use strict';

var core = require('@keystone-ui/core');
var React = require('react');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var button = require('@keystone-ui/button');
var AlertTriangleIcon = require('@keystone-ui/icons/icons/AlertTriangleIcon');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefault(_defineProperty);

class ErrorBoundary extends React.Component {
  constructor(...args) {
    super(...args);

    _defineProperty__default['default'](this, "state", {
      hasError: false,
      isReloading: false
    });

    _defineProperty__default['default'](this, "reloadPage", () => {
      this.setState({
        isReloading: true
      });
      window.location.reload();
    });
  }

  static getDerivedStateFromError(error) {
    return {
      error,
      hasError: true
    };
  }

  render() {
    if (this.state.hasError) {
      return core.jsx(ErrorContainer, null, core.jsx(core.Stack, {
        align: "center",
        gap: "medium"
      }, core.jsx(AlertTriangleIcon.AlertTriangleIcon, {
        size: "large"
      }), core.jsx("div", null, "Something went wrong."), core.jsx(button.Button, {
        size: "small",
        isLoading: this.state.isReloading,
        onClick: this.reloadPage
      }, "reload page")));
    }

    return this.props.children;
  }

}
const ErrorContainer = ({
  children
}) => {
  const {
    colors,
    shadow
  } = core.useTheme();
  return core.jsx(core.Center, {
    css: {
      minWidth: '100vw',
      minHeight: '100vh',
      backgroundColor: colors.backgroundMuted
    },
    rounding: "medium"
  }, core.jsx(core.Box, {
    css: {
      background: colors.background,
      boxShadow: shadow.s100
    },
    margin: "medium",
    padding: "xlarge",
    rounding: "medium"
  }, children));
};

exports.ErrorBoundary = ErrorBoundary;
exports.ErrorContainer = ErrorContainer;
